common/enclave/cc_data.cpp

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string>
+
+#include "base64.h"
+#include "cc_data.h"
+#include "error.h"
+#include "logging.h"
+
+#include <pb_decode.h>
+#include <pb_encode.h>
+#include "protos/fpc/attestation.pb.h"
+#include "protos/fpc/fpc.pb.h"
+
+#include "attestation-api/attestation/attestation.h"
+
+// ecc enclave global variable -- allocated dynamically
+cc_data* g_cc_data = NULL;
+
+bool cc_data::generate()
+{
+    return generate_keys();
+}
+
+bool cc_data::generate_keys()
+{
+    try
+    {
+        signature_key_.Generate();  // private key
+        verification_key_ = signature_key_.GetPublicKey();
+        decryption_key_.Generate();  // private key
+        encryption_key_ = decryption_key_.GetPublicKey();
+        cc_decryption_key_.Generate();  // private key
+        cc_encryption_key_ = cc_decryption_key_.GetPublicKey();
+
+        // debug
+        std::string s = verification_key_.Serialize();
+        LOG_DEBUG("enclave verification key: %s", s.c_str());
+    }
+    catch (...)
+    {
+        LOG_ERROR("error creating cryptographic keys");
+        return false;
+    }
+
+    return true;
+}
+
+bool cc_data::build_attested_data(ByteArray& attested_data)
+{
+    // estimate attested data size
+    // NOTE: the size is roughly estimated, it should be adapted
+    size_t attested_data_proto_size = attestation_parameters_.size() + cc_parameters_.size() +
+                                      host_parameters_.size() + (1 << 13);
+    uint8_t* attested_data_proto;
+    pb_ostream_t ostream;
+    bool b;
+
+    CATCH(b, attested_data.resize(attested_data_proto_size));
+    COND2ERR(!b);
+
+    attested_data_proto = attested_data.data();
+
+    ostream = pb_ostream_from_buffer(attested_data_proto, attested_data_proto_size);
+
+    {
+        // fpc_AttestedData_cc_params_tag
+        COND2ERR(!pb_encode_tag(&ostream, PB_WT_STRING, fpc_AttestedData_cc_params_tag));
+        COND2ERR(!pb_encode_string(
+            &ostream, (const unsigned char*)cc_parameters_.data(), cc_parameters_.size()));
+    }
+
+    {
+        // fpc_AttestedData_host_params_tag
+        COND2ERR(!pb_encode_tag(&ostream, PB_WT_STRING, fpc_AttestedData_host_params_tag));
+        COND2ERR(!pb_encode_string(
+            &ostream, (const unsigned char*)host_parameters_.data(), host_parameters_.size()));
+    }
+
+    {
+        // fpc_AttestedData_enclave_vk_tag
+        std::string s = verification_key_.Serialize();
+        COND2ERR(!pb_encode_tag(&ostream, PB_WT_STRING, fpc_AttestedData_enclave_vk_tag));
+        COND2ERR(!pb_encode_string(&ostream, (const unsigned char*)s.c_str(), s.length()));
+    }
+
+    // resize array to fit written data
+    attested_data.resize(ostream.bytes_written);
```

Comment:
Because the subfields are bytes or strings, and nanopb handles the encoding for these either through callbacks for each field or directly as I did. Same complexity anyway.

Reply:
but this you can handle via option files? and you do it only once instead of having to do it for each encode and decode? Anyway, not mutually exclusive but with option files the code gets easier to read. That of course is subjective ...

Reply:
It is possible to do that by specifying fixed size arrays -- doable, but you have to reserve enough size upfront.
Not sure you can do it otherwise.

Reply:
Ok, finally figured out my confusion: I was thinking of serializing/encoding a completely (deserialized) attested data object.  But as here we did not have to deserialize host-params and cc-params (yet), above code takes the already serialized version of them and then completes the serialization of just the attested data object itself.  In that case, you of course have to do a field-by-field insertion of the already serialized sub-field.

For marcus' benefit, as it is not immediately obvious from the code why this works (certainly was not to me ... :-)):
- the `PB_WT_STRING` is a tag not only for strings but also bytes, fixed-length-bytes and, importantly, sub-messages
- `pb_encode_submessage` does same encoding as `pb_encode_string`: put a length header and then the bytes of the string/(serialized) submessage